### PR TITLE
Unsupported server args

### PR DESCRIPTION
**Pull request by mpryc**
## Why the changes were made

Add unsupported args for the velero server via ConfigMap within the DPA namespace.

## How to test the changes made

1. Run the tests
2. Deploy oadp with this change, then

- Create ConfigMap with some name
- Inside ConfigMap create key/value pairs, where key will correspond to the velero server argument name and value is it's value
- Create DPA with annotation (replace <ConfigMap-Name> to adopt to your ConfigMap:
  `oadp.openshift.io/unsupported-velero-server-args=<ConfigMap-Name>`
- Verify the Velero server pod runs velero with the expected flags

Notes:
 - NonExisting ConfigMap referenced by the annotation will report error
 - Empty ConfigMap name will not cause override

### CODE REVIEW

File: controllers/velero.go [hunk 1]
```diff
@@ -144,6 +146,28 @@ func (r *DPAReconciler) veleroClusterRoleBinding(dpa *oadpv1alpha1.DataProtectio
...
+func (r *DPAReconciler) applyUnsupportedArgsOverride(dpa *oadpv1alpha1.DataProtectionApplication, veleroDeployment *appsv1.Deployment, veleroContainer *corev1.Container) error {
+	if configMapName, ok := dpa.Annotations[UnsupportedServerArgsAnnotation]; ok {
+		// Got an DPA annotation for Unsupported Server Args
+		if configMapName == "" {
+			// Can't be empty string, do not report error
+			return nil
+		}
+
+		unsupportedServerArgsCM := corev1.ConfigMap{}
+		if err := r.Get(r.Context, types.NamespacedName{Namespace: r.NamespacedName.Namespace, Name: configMapName}, &unsupportedServerArgsCM); err == nil {
```
Comment (mateusoliveira43) [inline controllers/velero.go:159]: would change order here to remove `else`

```go
if err := r.Get(r.Context, types.NamespacedName{Namespace: r.NamespacedName.Namespace, Name: configMapName}, &unsupportedServerArgsCM); err != nil {
    return err
}
r.Log.Info("Applying Unsupported Server Args.")
// if server args is set, override the default server args
veleroContainer.Args = common.GenerateCliArgsFromConfigMap("server", &unsupportedServerArgsCM)
```

File: controllers/velero.go
Comment (mateusoliveira43) [inline controllers/velero.go:1]: would be interesting (do not if it is possible) to add test with unsupported args in controllers/velero_test.go

Comment (mpryc) [reply controllers/velero.go:1]: yes on my TODO for 2morrow.

File: controllers/velero.go (hunk 1 above)
Comment (mpryc) [reply controllers/velero.go:159]: thanks, much simpler

File: controllers/velero.go
Comment (mpryc) [reply controllers/velero.go:1]: Added 2 tests in the last commit.

Comment (mateusoliveira43) [reply controllers/velero.go:1]: thanks

File: controllers/velero_test.go [hunk 2]
```diff
@@ -588,6 +588,117 @@ func TestDPAReconciler_buildVeleroDeployment(t *testing.T) {
...
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "test-Velero-CR",
+					Namespace: "test-ns",
+					Annotations: map[string]string{
+						"oadp.openshift.io/unsupported-velero-server-args": "unsupported-server-args-cm",
+					},
+				},
+				Spec: oadpv1alpha1.DataProtectionApplicationSpec{
+					Configuration: &oadpv1alpha1.ApplicationConfig{
+						Velero: &oadpv1alpha1.VeleroConfig{},
```
Comment (mateusoliveira43) [inline controllers/velero_test.go:612]: would add a new test, or do in this one, to confirm that any value user adds here is overwritten

File: controllers/velero.go [hunk 3]
```diff
@@ -45,6 +45,8 @@ const (
 
 	TrueVal  = "true"
 	FalseVal = "false"
+
+	UnsupportedServerArgsAnnotation = "oadp.openshift.io/unsupported-velero-server-args"
```
Comment (shubham-pampattiwar) [inline controllers/velero.go:49]: `s/UnsupportedServerArgsAnnotation/UnsupportedVeleroServerArgsAnnotation`

File: controllers/velero_test.go (hunk 2 above)
Comment (mpryc) [reply controllers/velero_test.go:612]: In the last commit added above test scenario and as well scenario where the ConfigMap name in the annotation was empty. I believe we now have really nice set of different tests for this feature.

File: controllers/velero.go (hunk 3 above)
Comment (mpryc) [reply controllers/velero.go]: Thanks, renamed in the last commit.

**Review by mateusoliveira43 — DISMISSED** [review]

File: controllers/velero.go [hunk 4]
```diff
@@ -144,6 +146,27 @@ func (r *DPAReconciler) veleroClusterRoleBinding(dpa *oadpv1alpha1.DataProtectio
...
+	if configMapName, ok := dpa.Annotations[UnsupportedVeleroServerArgsAnnotation]; ok {
+		// Got an DPA annotation for Unsupported Server Args
+		if configMapName == "" {
+			// Can't be empty string, do not report error
+			return nil
+		}
+
+		unsupportedServerArgsCM := corev1.ConfigMap{}
+		if err := r.Get(r.Context, types.NamespacedName{Namespace: dpa.Namespace, Name: configMapName}, &unsupportedServerArgsCM); err != nil {
+			return err
```
Comment (shubham-pampattiwar) [inline controllers/velero.go:160]: For non-existing config map I get the following error:
```
  conditions:
  - lastTransitionTime: "2024-06-21T16:18:21Z"
    message: ConfigMap "non-existing-cm" not found
    reason: Error
    status: "False"
    type: Reconciled

```

Should we add more color to the error message ? maybe something like `Config <name> for unsupported velero server args not found` ? Current error also looks good though, up to you.

**Comment by shubham-pampattiwar** [comment]
For existing config map with no quotes in values the default cases did not work for me:
```
shubham@fedora ~/oadp-operator -  (mpryc-unsup-server) $ oc get po
NAME                                                              READY   STATUS      RESTARTS      AGE
0d14697ee7cbf27b228781c31c6c141a9225f4681ed99252ab37a0c70bx6hwc   0/1     Completed   0             9m51s
node-agent-4bk82                                                  1/1     Running     0             48s
node-agent-f8flz                                                  1/1     Running     0             48s
openshift-adp-controller-manager-759544c68f-wdm22                 1/1     Running     0             9m33s
ttl-sh-oadp-operator-bundle-2181163-1h                            1/1     Running     0             10m
velero-7c585cdf5-pt922                                            0/1     Error       3 (30s ago)   48s
shubham@fedora ~/oadp-operator -  (mpryc-unsup-server) $ oc logs velero-7c585cdf5-pt922
Defaulted container "velero" out of: velero, openshift-velero-plugin (init), velero-plugin-for-aws (init)
time="2024-06-21T16:37:55Z" level=info msg="setting log-level to INFO" logSource="pkg/cmd/server/server.go:199"
time="2024-06-21T16:37:55Z" level=info msg="Starting Velero server konveyor-dev (-)" logSource="pkg/cmd/server/server.go:201"
time="2024-06-21T16:37:55Z" level=info msg="1 feature flags enabled ['EnableCSI']" logSource="pkg/cmd/server/server.go:203"
An error occurred: invalid uploader type ''kopia'', valid upload types are: 'restic', 'kopia'
shubham@fedora ~/oadp-operator -  (mpryc-unsup-server) $ oc get cm my-config -oyaml
apiVersion: v1
data:
  features: EnableCSI
  uploader-type: kopia
kind: ConfigMap
metadata:
  creationTimestamp: "2024-06-21T16:33:01Z"
  name: my-config
  namespace: openshift-adp
  resourceVersion: "99653500"
  uid: c1e32569-966c-457b-8725-2b6092dd74c3
shubham@fedora ~/oadp-operator -  (mpryc-unsup-server) $
```

**Comment by mpryc** [comment]
> For existing config map with no quotes in values the default cases did not work for me:
> 
> ```
> shubham@fedora ~/oadp-operator -  (mpryc-unsup-server) $ oc get po
> NAME                                                              READY   STATUS      RESTARTS      AGE
> 0d14697ee7cbf27b228781c31c6c141a9225f4681ed99252ab37a0c70bx6hwc   0/1     Completed   0             9m51s
> node-agent-4bk82                                                  1/1     Running     0             48s
> node-agent-f8flz                                                  1/1     Running     0             48s
> openshift-adp-controller-manager-759544c68f-wdm22                 1/1     Running     0             9m33s
> ttl-sh-oadp-operator-bundle-2181163-1h                            1/1     Running     0             10m
> velero-7c585cdf5-pt922                                            0/1     Error       3 (30s ago)   48s
> shubham@fedora ~/oadp-operator -  (mpryc-unsup-server) $ oc logs velero-7c585cdf5-pt922
> Defaulted container "velero" out of: velero, openshift-velero-plugin (init), velero-plugin-for-aws (init)
> time="2024-06-21T16:37:55Z" level=info msg="setting log-level to INFO" logSource="pkg/cmd/server/server.go:199"
> time="2024-06-21T16:37:55Z" level=info msg="Starting Velero server konveyor-dev (-)" logSource="pkg/cmd/server/server.go:201"
> time="2024-06-21T16:37:55Z" level=info msg="1 feature flags enabled ['EnableCSI']" logSource="pkg/cmd/server/server.go:203"
> An error occurred: invalid uploader type ''kopia'', valid upload types are: 'restic', 'kopia'
> shubham@fedora ~/oadp-operator -  (mpryc-unsup-server) $ oc get cm my-config -oyaml
> apiVersion: v1
> data:
>   features: EnableCSI
>   uploader-type: kopia
> kind: ConfigMap
> metadata:
>   creationTimestamp: "2024-06-21T16:33:01Z"
>   name: my-config
>   namespace: openshift-adp
>   resourceVersion: "99653500"
>   uid: c1e32569-966c-457b-8725-2b6092dd74c3
> shubham@fedora ~/oadp-operator -  (mpryc-unsup-server) $
> ```

Ok so it seems like the `--uploader-type='kopia'` is an invalid parameter to be passed. Maybe we should *not* surround any value at all?

**Comment by mpryc** [comment]
@shubham-pampattiwar ok so I think we need to adjust slightly the way design is defined and implementation.

The problem is with how the args are later passed to the resulting velero Pod->Container spec and the problem is that the args are considered strings, so we have problems with wrapping one string into another.


```yaml
      image: >-
        registry.redhat.io/oadp/oadp-velero-rhel9@sha256:06482afcea65eff184901c63cbe9fc6e5b3172d23857301ad4e7b4daf362e79c
      args:
        - server
        - '--uploader-type=restic'
        - '--fs-backup-timeout=1h'
        - >-
          --restore-resource-priorities=securitycontextconstraints,customresourcedefinitions,namespaces,managedcluster.cluster.open-cluster-management.io,managedcluster.clusterview.open-cluster-management.io,klusterletaddonconfig.agent.open-cluster-management.io,managedclusteraddon.addon.open-cluster-management.io,storageclasses,volumesnapshotclass.snapshot.storage.k8s.io,volumesnapshotcontents.snapshot.storage.k8s.io,volumesnapshots.snapshot.storage.k8s.io,datauploads.velero.io,persistentvolumes,persistentvolumeclaims,serviceaccounts,secrets,configmaps,limitranges,pods,replicasets.apps,clusterclasses.cluster.x-k8s.io,endpoints,services,-,clusterbootstraps.run.tanzu.vmware.com,clusters.cluster.x-k8s.io,clusterresourcesets.addons.cluster.x-k8s.io

```

Could you please paste here your broken velero Pod specification?

**Comment by shubham-pampattiwar** [comment]
@mpryc Here you go: 
```
shubham@fedora ~/oadp-operator -  (mpryc-unsup-server) $ oc get po velero-7c585cdf5-7kkkn -oyaml | grep args -7
    kind: ReplicaSet
    name: velero-7c585cdf5
    uid: c0a30d53-227c-4aa4-92dc-7670d9c06329
  resourceVersion: "101617132"
  uid: fb136af6-2c84-4a7e-802a-8d0b1744def9
spec:
  containers:
  - args:
    - server
    - --features='EnableCSI'
    - --uploader-type='kopia'
    command:
    - /velero
    env:
    - name: VELERO_SCRATCH_DIR

```

**Comment by mpryc** [comment]
> @mpryc Here you go:
> 
> ```
> shubham@fedora ~/oadp-operator -  (mpryc-unsup-server) $ oc get po velero-7c585cdf5-7kkkn -oyaml | grep args -7
>     kind: ReplicaSet
>     name: velero-7c585cdf5
>     uid: c0a30d53-227c-4aa4-92dc-7670d9c06329
>   resourceVersion: "101617132"
>   uid: fb136af6-2c84-4a7e-802a-8d0b1744def9
> spec:
>   containers:
>   - args:
>     - server
>     - --features='EnableCSI'
>     - --uploader-type='kopia'
>     command:
>     - /velero
>     env:
>     - name: VELERO_SCRATCH_DIR
> ```

Interesting, ok so we need to modify that part, because it should be more of ` - "--uploader-type='kopia'"` or ` - '--uploader-type="kopia"'` .

I wonder how this behaves when it's the string of a format without `'`, so:
```yaml
    kind: ReplicaSet
    name: velero-7c585cdf5
    uid: c0a30d53-227c-4aa4-92dc-7670d9c06329
  resourceVersion: "101617132"
  uid: fb136af6-2c84-4a7e-802a-8d0b1744def9
spec:
  containers:
  - args:
    - server
    - --features=EnableCSI
    - --uploader-type=kopia
    command:
    - /velero
    env:
    - name: VELERO_SCRATCH_DIR
 ```

**Comment by shubham-pampattiwar** [comment]
@mpryc Yeah I think we need the command without any quotes:
```
    kind: ReplicaSet
    name: velero-7c585cdf5
    uid: c0a30d53-227c-4aa4-92dc-7670d9c06329
  resourceVersion: "101617132"
  uid: fb136af6-2c84-4a7e-802a-8d0b1744def9
spec:
  containers:
  - args:
    - server
    - --features=EnableCSI
    - --uploader-type=kopia
    command:
    - /velero
    env:
    - name: VELERO_SCRATCH_DIR
```

**Comment by mpryc** [comment]
@shubham-pampattiwar rebased on a modified design/function and I think it's now correctly handling your ConfigMap as well other args passed. Please review.

File: controllers/velero.go (hunk 4 above)
Comment (mpryc) [reply controllers/velero.go:160]: I prefer not to mess up with standard error messages if they are good enough, so I will leave it.

**Comment by mateusoliveira43** [comment]
/hold
review #1423 first

**Comment by mpryc** [comment]
Squashed all the commits into one and rebased on top of master which now includes #1423. Should be easier to cherry pick to other branch and it's easier to review.

**Comment by shubham-pampattiwar** [comment]
PR worked for me with the latest changes:
```
shubham@fedora ~/oadp-operator -  (mpryc-server-args) $ oc get cm my-config -oyaml
apiVersion: v1
data:
  features: EnableCSI
  fs-backup-timeout: 10h
  uploader-type: kopia
kind: ConfigMap
metadata:
  creationTimestamp: "2024-07-01T23:07:27Z"
  name: my-config
  namespace: openshift-adp
  resourceVersion: "106776916"
  uid: 127a5904-da3f-4917-9483-f8fd177aba7b
shubham@fedora ~/oadp-operator -  (mpryc-server-args) $ oc get po velero-7b8c5dc7df-jr4s9 -oyaml | grep args -7
    kind: ReplicaSet
    name: velero-7b8c5dc7df
    uid: 3f20ea1e-9959-44d2-bf5b-e2ebdc2abe92
  resourceVersion: "106777116"
  uid: 688f8716-e0c0-4787-977c-0d66391acb95
spec:
  containers:
  - args:
    - server
    - --features=EnableCSI
    - --fs-backup-timeout=10h
    - --uploader-type=kopia
    command:
    - /velero
    env:
shubham@fedora ~/oadp-operator -  (mpryc-server-args)
```

**Review by shubham-pampattiwar — APPROVED** [review]
/lgtm

**Pull request merged**